src/validator/IndexScanValidator.cpp

```diff
@@ -93,26 +95,147 @@ Status IndexScanValidator::prepareFilter() {
     }
 
     auto *filter = sentence->whereClause()->filter();
-    auto ret = checkFilter(filter, *sentence->from());
-    NG_RETURN_IF_ERROR(ret);
     storage::cpp2::IndexQueryContext ctx;
-    ctx.set_filter(Expression::encode(*filter));
+    if (needTextSearch(filter)) {
+        NG_RETURN_IF_ERROR(checkTSService());
+        if (!textSearchReady_) {
+            return Status::Error("Text search service not ready");
+        }
+        auto retExpr = rewriteTSFilter(filter);
+        if (!retExpr.ok()) {
+            return retExpr.status();
+        }
+        if (isEmptyResultSet_) {
+            // return empty result direct.
+            return Status::OK();
+        }
+        auto ret = checkFilter(retExpr.value());
+        NG_RETURN_IF_ERROR(ret);
```

Comment:
Leak retExpr.value() memory when fail to checkFilter

Reply:
Good point, fixed . thanks.